File: Blind 75/16 - Intervals/253-meeting-rooms-II.py
'''
*LeetCode premium problem
'''

from time import time
from typing import List, Tuple


class Solution:
    def minMeetingRooms(self, intervals: List[Tuple[int, int]]) -> int:
        start_times = sorted([start for start, _ in intervals])
        end_times = sorted([end for _, end in intervals])
        res, count = 0, 0

        # Accumulate results with two pointers
        i, j = 0, 0
        while i < len(intervals):
            if start_times[i] < end_times[j]:
                # Number of meetings has increased
                i += 1
                count += 1
            else:
                # Number of meetings has decreased
                j += 1
                count -= 1
            res = max(res, count)

        return res

    def reference(self, intervals: List[Tuple[int, int]]) -> int:
        start = sorted([i[0] for i in intervals])
        end = sorted([i[1] for i in intervals])

        res, count = 0, 0
        s, e = 0, 0
        while s < len(intervals):
            if start[s] < end[e]:
                s += 1
                count += 1
            else:
                e += 1
                count -= 1
            res = max(res, count)
        return res

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.minMeetingRooms(case))
                else:
                    self.minMeetingRooms(case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        [(0, 30), (5, 10), (15, 20)],
        # Additional
        [(5, 10), (15, 20), (0, 30)],
        [(0, 1)],
        [],
    ]
    test.quantify(test_cases)

File: Blind 75/16 - Intervals/JavaScript/253-meeting-rooms-II.js
const { performance } = require("perf_hooks");

class Solution {
  method(intervals) {
    intervals.sort((a, b) => a[0] - b[0]);
    let count = 1;
    for (let i = 1; i < intervals.length; i++) {
      if (intervals[i - 1][1] > intervals[i][0]) count += 1;
    }

    return count;
  }

  reference(intervals) {
    const { start, end } = this.splitIntervals(intervals);
    let [minRooms, startIndex, endIndex] = [0, 0, 0];

    while (startIndex < intervals.length) {
      const [currStart, prevEnd] = [start[startIndex], end[endIndex]];

      const hasGap = prevEnd <= currStart;
      if (hasGap) {
        minRooms--;
        endIndex++;
      }

      minRooms++;
      startIndex++;
    }

    return minRooms;
  }

  splitIntervals = (intervals, start = [], end = []) => {
    for (const [startTime, endTime] of intervals) {
      start.push(startTime);
      end.push(endTime);
    }

    const comparator = (a, b) => a - b;

    start.sort(comparator);
    end.sort(comparator);

    return { start, end };
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [
    [0, 30],
    [5, 10],
    [15, 20],
  ],
  // Additional
  [
    [5, 10],
    [15, 20],
    [0, 30],
  ],
];
test.quantify(testCases);
